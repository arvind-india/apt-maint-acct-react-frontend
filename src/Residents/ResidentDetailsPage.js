import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import Select from 'react-select';

import {
    Form,
    Button,
//    FormGroup,
    FormText,
    Input,
    Label,
} from 'reactstrap'

import { residentActions as actions, alertActions, userActions } from '../_actions'
import { RESIDENT_TYPES } from '../_constants'
import { FlashMessage } from '../_components'

let module = 'residents' // module name

export class ResidentDetails extends React.Component {

  constructor(props) {
    super(props)
    const { location } = props
    let model = location.state.model // model supplied from list page
//    let ops = model.operations?model.operations:''
    let initializeModel = {
      id: model.id,
      owner_id: model.owner_id,
      first_name: model.first_name,
      last_name: model.last_name,
      is_a: model.is_a,
      occupied_on: model.occupied_on,
      vacated_on: model.vacated_on
    }
    this.state = {
      model: initializeModel,           // model to edit
      submitted: false,
      adding: model.id === 0
    }
    this.handleChange = this.handleChange.bind(this)
    this.handleUserNameChange = this.handleUserNameChange.bind(this)
    this.handleResidentTypeChange = this.handleResidentTypeChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    //dispatch(alertActions.clear())  // clear alert messages from other pages
    //this.props.clearAlert()
  }

  componentDidMount() {
    //console.log('state now: ', this.state)
    //this.props.dispatch(userActions.getAll())
    //this.props.getAllUser()
  }

  canBeSaved() { // check for changes in model, if changes present, it can save
    const { model } = this.state
    if (!model.owner_id) return false
    if (!model.first_name) return false
    if (!model.last_name) return false
    if (!model.is_a) return false
    return true // can save changes
  }

  render() {
    const { alert, users } = this.props
    return (
      <div>
        <h2>Resident Details</h2>
        {alert.message && <FlashMessage text={alert.message} color={alert.color} delay={2100}/>}
        {this.validateForm()}
        {users.items && this.show()}
      </div>
    )
  }
  validateForm() {
    if ( this.changedProps().length === 0 ) {
      this.validationMsg = 'No changes to save'
      this.formValid = false
      return null
    }
    if( this.canBeSaved() ) {
      this.validationMsg = 'Save Changes'
      this.formValid = true
      return null
    }
    // finally, if reached here
    this.validationMsg = 'Missing "Required Data"...'
    this.formValid = false
  }

  show(){
    const { adding } = this.state
    const { authzn } = this.props
    let title = adding?'Add':authzn.allowsEdit?'Edit':'View'
    return <Form id="residentDetailsForm" onSubmit={this.handleSubmit} className="grid-form">
      <fieldset>
  			<legend>{title}</legend>
        <div data-row-span="2">
          {this.showUserName()}
          {this.showResidentTypeIsA()}
        </div>
        <div data-row-span="2">
          {this.showFirstName()}
          {this.showLastName()}
        </div>
        <div data-row-span="2">
          {this.showOccupiedOn()}
          {this.showVacatedOn()}
        </div>
      </fieldset>
      <br/>
      <Button
        type="submit"
        color="primary"
        disabled={!this.formValid}
        hidden={!authzn.allowsEdit}
        title={this.validationMsg}
        >Save</Button>
      <Button
        color="link"
        ><Link
          to="/residents"
          className="text-danger"
          title="Go to Residents"
          >Cancel</Link></Button>
    </Form>
  }

  handleSubmit(event) {
    const { model } = this.state
    //const { dispatch } = this.props

    event.preventDefault()
    this.setState({ submitted: true })
    this.props.saveChanges(model)
/*
    if ( this.changedProps().length === 0 ) {
      dispatch(alertActions.error('No changes found...'))
    } else if(this.canBeSaved()){
console.log('Resident model to be saved: ', model)
      dispatch(actions.saveChanges(model))
    } else {
      dispatch(alertActions.error('Missing data'))
    }
*/
  }

  changedProps() {
    const { model } = this.state
    const { location } = this.props
    let modelDB = location.state.model
    let props = []

    // check for changes in model props
    for(const prop in model) {
      if( prop === 'id') continue // exclude 'id' from comparision
      if(modelDB[prop] !== model[prop]) { // if data is changed wrt data in database
        props.push(prop)
      }
    }
    return props;
  }

  showUserName() {
    const { submitted, model } = this.state
    const { users } = this.props
    return <div data-field-span="1">
				<Label>User Name</Label>
        <Select
          id="ownerId"
          name="owner_id"
          value={model.owner_id}
          multi={false}
          joinValues={false}
          simpleValue={true}
          placeholder="Select User name..."
          onChange={this.handleUserNameChange}
          valueKey="id"
          labelKey="name"
          options={users.items}
        />
        {submitted && !model.owner_id
          && <FormText color="danger">User Name is required</FormText>}
			</div>
  }
  handleUserNameChange(selectedOption) {
    const { model } = this.state
    this.setState({
      model: {
        ...model,
        owner_id: selectedOption
      }
    })
  }
  showFirstName() {
    const { submitted, model } = this.state
    //console.log('model is:..... ', model)
    return <div data-field-span="1">
				<Label>First Name</Label>
        <Input
          id="firstName"
          type="text"
          name="first_name"
          value={model.first_name}
          placeholder="First Name here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.first_name
          && <FormText color="danger">First Name is required</FormText>}
			</div>
  }

  handleChange(event) {
    const { name, value } = event.target
    const { model } = this.state
    this.setState({
        model: {
          ...model,
          [name]: value
        }
    })
  }

  showLastName() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Last Name</Label>
        <Input
          id="lastName"
          type="text"
          name="last_name"
          value={model.last_name}
          placeholder="Last Name here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.last_name
          && <FormText color="danger">Last Name is required</FormText>}
			</div>
  }
  showResidentTypeIsA() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Resident Type</Label>
        <Select
          id="isA"
          name="is_a"
          value={model.is_a}
          multi={false}
          joinValues={false}
          simpleValue={true}
          placeholder="Select Resident Type..."
          onChange={this.handleResidentTypeChange}
          valueKey="name"
          labelKey="label"
          options={RESIDENT_TYPES}
        />
        {submitted && !model.is_a
          && <FormText color="danger">Resident Type is required</FormText>}
			</div>
  }
  handleResidentTypeChange(selectedOption) {
    const { model } = this.state
    this.setState({
      model: {
        ...model,
        is_a: selectedOption
      }
    })
  }

  showOccupiedOn() {
    const { model } = this.state
    return <div data-field-span="1">
				<Label>Occupied On</Label>
        <Input
          id="occupiedOn"
          type="date"
          name="occupied_on"
          value={model.occupied_on}
          placeholder="Date of occupation of premise"
          className="inputField"
          onChange={this.handleChange}
        />
			</div>
  }
  showVacatedOn() {
    const { model } = this.state
    return <div data-field-span="1">
				<Label>Vacated On</Label>
        <Input
          id="vacatedOn"
          type="date"
          name="vacated_on"
          value={model.vacated_on}
          placeholder="Date of vacation of premise"
          className="inputField"
          onChange={this.handleChange}
        />
			</div>
  }

}

function mapStateToProps(state) {
  const { alert, authorizations, users } = state
//  const { user } = authentication
  const authzn = authorizations[module]
  return {
//    user,
    alert,
    authzn,
    users
  }
}

function mapDispatchToProps(dispatch) {
  return {
/*    clearAlert: () => {
      dispatch(alertActions.clear())
    }, */
    saveChanges: (model) => {
      dispatch(actions.saveChanges(model))
    },
    error: (msg) => {
      dispatch(alertActions.error(msg))
    },
    getAllUser: () => {
      dispatch(userActions.getAll())
    }
  }
}

//const connectedDetailsPage = connect(mapStateToProps)(ResidentDetailsPage)
//export { connectedDetailsPage as ResidentDetailsPage }
export default connect(mapStateToProps, mapDispatchToProps)(ResidentDetails)
