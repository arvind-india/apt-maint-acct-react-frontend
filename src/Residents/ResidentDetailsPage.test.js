// ResidentsPage.test.js0]

import React from 'react'
import { shallow } from 'enzyme'

import { ResidentDetails } from './ResidentDetailsPage'

const mockResidents = [
  {
    id: 1,
    owner_id: 'guest',
    first_name: 'guest',
    last_name: 'user',
    is_a: 'guest',
    occupied_on: '2018-03-31',
    vacated_on: '2020-03-31'
  },
  {
    id: 2,
    owner_id: 'og1',
    first_name: 'g1',
    last_name: 'user',
    is_a: 'owner',
    occupied_on: '2018-04-31',
    vacated_on: '2020-04-31'
  }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    residents: {
      items: []
    },
    users: {
      items: [
        { name: 'guest' },
        { name: 'og1' },
        { name: 'tg1' }
      ]
    }
  },
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    residents: {
      items: []
    },
    users: {
      items: [
        { name: 'guest' },
        { name: 'og1' },
        { name: 'tg1' }
      ]
    }
  }
]
const mockLocations = [
  {
    state: {
      model: mockResidents[0],
      submitted: false,
      adding: false
    }
  },
  {
    state: {
      model: mockResidents[1],
      submitted: false,
      adding: false
    }
  }
]

const mocks = [
  {
    ownerId: mockResidents[0].owner_id,
    isA: mockResidents[0].is_a,
    firstName: {name: 'first_name', value: mockResidents[0].first_name},
    lastName: {name: 'last_name', value: mockResidents[0].last_name},
    occupiedOn: {name: 'occupied_on', value: mockResidents[0].occupied_on},
    vacatedOn: {name: 'vacated_on', value: mockResidents[0].vacated_on}
  },
  {
    ownerId: mockResidents[1].owner_id,
    isA: mockResidents[1].is_a,
    firstName: {name: 'first_name', value: mockResidents[1].first_name},
    lastName: {name: 'last_name', value: mockResidents[1].last_name},
    occupiedOn: {name: 'occupied_on', value: mockResidents[1].occupied_on},
    vacatedOn: {name: 'vacated_on', value: mockResidents[1].vacated_on}  }
]

const component = shallow(<ResidentDetails {...mockProps[0]} location={mockLocations[0]}/>)

describe('ResidentsPage params test', () => {
  it('should render ResidentsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('model')).toBeDefined()
  })
  it('should have submitted state', () => {
    expect(component.state('submitted')).toBeDefined()
  })
  it('should have adding state', () => {
    expect(component.state('adding')).toBeDefined()
  })
})


describe('ResidentDetailsPage events test 1', () => {
  it('should call clearAlert on submit and no crash', () => {
    component.find('#residentDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].clearAlert.mock.calls.length).toEqual(1)
  })

  it('should be called with required states as arguments', () => {
    component.find('#ownerId').simulate('change', mocks[0].ownerId)
    component.find('#isA').simulate('change', mocks[0].isA)
    component.find('#firstName').simulate('change', {target: mocks[0].firstName}) // calling again to toggle from '' to 'R'
    component.find('#lastName').simulate('change', {target: mocks[0].lastName})
    component.find('#occupiedOn').simulate('change', {target: mocks[0].occupiedOn})
    component.find('#vacatedOn').simulate('change', {target: mocks[0].vacatedOn})
    component.find('#residentDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].saveChanges.mock.calls[1][0]).toEqual(mockResidents[0])
  })

})


const component2 = shallow(<ResidentDetails {...mockProps[1]} location={mockLocations[1]}/>)
describe('ResidentDetailsPage events test 2', () => {
  it('should be called with required states as arguments', () => {
    component2.find('#ownerId').simulate('change', mocks[1].ownerId)
    component2.find('#isA').simulate('change', mocks[1].isA)
    component2.find('#firstName').simulate('change', {target: mocks[1].firstName}) // calling again to toggle from '' to 'R'
    component2.find('#lastName').simulate('change', {target: mocks[1].lastName})
    component2.find('#occupiedOn').simulate('change', {target: mocks[1].occupiedOn})
    component2.find('#vacatedOn').simulate('change', {target: mocks[1].vacatedOn})
    component2.find('#residentDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[1].saveChanges.mock.calls[0][0]).toEqual(mockResidents[1])
  })
})
