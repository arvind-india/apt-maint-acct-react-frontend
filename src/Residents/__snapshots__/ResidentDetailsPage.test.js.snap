// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResidentsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h2>
        Resident Details
    </h2>
    <FlashMessage
        color={undefined}
        delay={2100}
        text="testingAlert"
    />
    <Form
        className="grid-form"
        id="residentDetailsForm"
        onSubmit={[Function]}
        tag="form"
    >
        <fieldset>
            <legend>
                Edit
            </legend>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        User Name
                    </Label>
                    <Select
                        arrowRenderer={[Function]}
                        autosize={true}
                        backspaceRemoves={true}
                        backspaceToRemoveMessage="Press backspace to remove {label}"
                        clearAllText="Clear all"
                        clearRenderer={[Function]}
                        clearValueText="Clear value"
                        clearable={true}
                        closeOnSelect={true}
                        deleteRemoves={true}
                        delimiter=","
                        disabled={false}
                        escapeClearsValue={true}
                        filterOptions={[Function]}
                        id="ownerId"
                        ignoreAccents={true}
                        ignoreCase={true}
                        inputProps={Object {}}
                        isLoading={false}
                        joinValues={false}
                        labelKey="name"
                        matchPos="any"
                        matchProp="any"
                        menuBuffer={0}
                        menuRenderer={[Function]}
                        multi={false}
                        name="owner_id"
                        noResultsText="No results found"
                        onBlurResetsInput={true}
                        onChange={[Function]}
                        onCloseResetsInput={true}
                        onSelectResetsInput={true}
                        openOnClick={true}
                        optionComponent={[Function]}
                        options={
                            Array [
                                Object {
                                  "name": "guest",
                                },
                                Object {
                                  "name": "og1",
                                },
                                Object {
                                  "name": "tg1",
                                },
                              ]
                        }
                        pageSize={5}
                        placeholder="Select User name..."
                        removeSelected={true}
                        required={false}
                        rtl={false}
                        scrollMenuIntoView={true}
                        searchable={true}
                        simpleValue={true}
                        tabSelectsValue={true}
                        trimFilter={true}
                        value="guest"
                        valueComponent={[Function]}
                        valueKey="id"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Resident Type
                    </Label>
                    <Select
                        arrowRenderer={[Function]}
                        autosize={true}
                        backspaceRemoves={true}
                        backspaceToRemoveMessage="Press backspace to remove {label}"
                        clearAllText="Clear all"
                        clearRenderer={[Function]}
                        clearValueText="Clear value"
                        clearable={true}
                        closeOnSelect={true}
                        deleteRemoves={true}
                        delimiter=","
                        disabled={false}
                        escapeClearsValue={true}
                        filterOptions={[Function]}
                        id="isA"
                        ignoreAccents={true}
                        ignoreCase={true}
                        inputProps={Object {}}
                        isLoading={false}
                        joinValues={false}
                        labelKey="label"
                        matchPos="any"
                        matchProp="any"
                        menuBuffer={0}
                        menuRenderer={[Function]}
                        multi={false}
                        name="is_a"
                        noResultsText="No results found"
                        onBlurResetsInput={true}
                        onChange={[Function]}
                        onCloseResetsInput={true}
                        onSelectResetsInput={true}
                        openOnClick={true}
                        optionComponent={[Function]}
                        options={
                            Array [
                                Object {
                                  "label": "Owner",
                                  "name": "owner",
                                },
                                Object {
                                  "label": "Tenant",
                                  "name": "tenant",
                                },
                                Object {
                                  "label": "Relative",
                                  "name": "relative",
                                },
                                Object {
                                  "label": "Friend",
                                  "name": "friend",
                                },
                                Object {
                                  "label": "Guest",
                                  "name": "guest",
                                },
                              ]
                        }
                        pageSize={5}
                        placeholder="Select Resident Type..."
                        removeSelected={true}
                        required={false}
                        rtl={false}
                        scrollMenuIntoView={true}
                        searchable={true}
                        simpleValue={true}
                        tabSelectsValue={true}
                        trimFilter={true}
                        value="guest"
                        valueComponent={[Function]}
                        valueKey="name"
                    />
                </div>
            </div>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        First Name
                    </Label>
                    <Input
                        className="inputField"
                        id="firstName"
                        name="first_name"
                        onChange={[Function]}
                        placeholder="First Name here"
                        tag="p"
                        type="text"
                        value="guest"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Last Name
                    </Label>
                    <Input
                        className="inputField"
                        id="lastName"
                        name="last_name"
                        onChange={[Function]}
                        placeholder="Last Name here"
                        tag="p"
                        type="text"
                        value="user"
                    />
                </div>
            </div>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Occupied On
                    </Label>
                    <Input
                        className="inputField"
                        id="occupiedOn"
                        name="occupied_on"
                        onChange={[Function]}
                        placeholder="Date of occupation of premise"
                        tag="p"
                        type="date"
                        value="2018-03-31"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Vacated On
                    </Label>
                    <Input
                        className="inputField"
                        id="vacatedOn"
                        name="vacated_on"
                        onChange={[Function]}
                        placeholder="Date of vacation of premise"
                        tag="p"
                        type="date"
                        value="2020-03-31"
                    />
                </div>
            </div>
        </fieldset>
        <br />
        <Button
            color="primary"
            disabled={true}
            hidden={false}
            tag="button"
            title="No changes to save"
            type="submit"
        >
            Save
        </Button>
        <Button
            color="link"
            tag="button"
        >
            <Link
                className="text-danger"
                replace={false}
                title="Go to Residents"
                to="/residents"
            >
                Cancel
            </Link>
        </Button>
    </Form>
</div>,
]
`;
